src/multi_processing/generator.py

```diff
@@ -0,0 +1,151 @@
+import time
+import multi_processing.pool_tasks as mp_p
+
+
+class Generator():
+    """ A class to coordinate the generation of domain objects as per jobs
+    observed within a multiprocessed queue. Data is then generated in a pool
+    before the result of which is put into the write queue.
+
+    Attributes
+    ----------
+    generate_queue : Multiprocess Queue
+        Multiprocess safe queue from which jobs to generate data is taken
+    write_queue : Multiprocess Queue
+        Multiprocess safe queue on which generated data is placed
+    terminate : Boolean
+        Boolean flag which when True indicates the coordinator is to terminate
+
+    Methods
+    -------
+    start(obj_class, pool_size)
+        Begin the cycle of waiting for, formatting and running jobs. Continue
+        this until termination instruction observed
+    wait_for_jobs()
+        Sleep whilst the generation queue is empty
+    format_jobs(obj_class, pool_size)
+        Retrieve jobs from queue and place them into an iterable structure
+    issue_termination()
+        Set flag to terminate to True
+    run_jobs(job_list, pool_size)
+        Pass formatted jobs to the generation pool for execution
+    get_generation_queue()
+        Return a pointer to the multiprocess safe generation job queue
+    """
+
+    def __init__(self, generate_queue, write_queue):
+        """ Assign variables from input, and set termination to False
+
+        Parameters
+        ----------
+        generate_queue : Multiprocessed Queue
+            Queue containing jobs to generate an amount of domain objects
+        write_queue : Multiprocessed Queue
+            Queue containing resultant generated objects for writing to file
+        """
+
+        self.generate_queue = generate_queue
+        self.write_queue = write_queue
+        self.terminate = False
+
+    def start(self, obj_class, pool_size):
+        """ Begin the cycle of waiting for, formatting, and running jobs,
+        continuing this until an instruction to terminate is observed.
+
+        Parameters
+        ----------
+        obj_class : String
+            Instantiated domain object for calling .generate over
+        pool_size : int
+            The number of processes running in the generator's pool
+        """
+
+        while not self.terminate:
+            self.wait_for_jobs()
+            job_list = self.format_jobs(obj_class, pool_size)
+            self.run_jobs(job_list, pool_size)
+
+        self.write_queue.put("terminate")
+
+    def wait_for_jobs(self):
+        """ Sleep until jobs are on the queue """
+
+        while self.generate_queue.empty():
+            time.sleep(1)
+        return
+
+    def format_jobs(self, obj_class, pool_size):
+        """ Takes jobs from the Multiprocess Safe Queue and places them into
```

Comment:
confused. Why is the job_list < poolsize*2 the signal to return the job list? So it looks like the jobs aren't streamed, they're batched in sizes of 2*poolsize. 

Also, it looks like we producer batches of 2*poolsize and the run 'start' one more time for any leftover. Is that correct? seems a bit messy. Not that I have a better suggestion off the top of my head but i would think there is a cleaner way to do this